posts/2021-05-07-caught-red-handed.md

```diff
@@ -0,0 +1,125 @@
+[pr-83007]: https://github.com/rust-lang/rust/pull/83007
+
+When we first turned on `verify-ich` by default, we assumed we would have time to iron out all the known causes of unstable fingerprints. When we realized this assumption was false, we started [making plans][issue-84970] to improve the user-experience, so that the diagnostic issued by the check would do a better job of telling the programmer what to do in response. But we made a mistake: We thought that the switch for this check was not going to be on the Rust stable channel until version 1.53.
+
+[issue-84970]: https://github.com/rust-lang/rust/issues/84970
+
+It turns out `verify-ich` was turned on in version 1.52.0, which was [released recently][].
+
+[released recently]: /2021/05/06/Rust-1.52.0.html
+
+## How does this show up
+
+
+Essentially, for some crates, certain sequences of edit-compile cycles will cause `rustc` to hit the "unstable fingerprints" ICE. I showed one example at the start of this blog post.
+
+Another recent example looks [like this](https://github.com/rust-lang/rust/issues/85039):
+
+```
+thread 'rustc' panicked at 'found unstable fingerprints for predicates_of(<massive text describing rustc internals elided>)', /rustc/88f19c6dab716c6281af7602e30f413e809c5974/compiler/rustc_query_system/src/query/plumbing.rs:593:5
+```
+
+They all arise from inconsistencies when comparing the incremental-compilation cache stored on disk against the values computed during a current `rustc` invocation, which means they all arise from using incremental compilation.
+
+There are three ways that you may have incremental compilation turned on: You may have set the [environment variable][env-vars] `CARGO_INCREMENTAL=1`, or you may have enabled the `build.incremental` [setting in your Cargo.toml][cargo-toml], or you may be building with the `dev` or `test` [profiles][], which default to having incremental compilation enabled.
+
+[env-vars]: https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-reads
+[cargo-toml]: https://doc.rust-lang.org/cargo/reference/config.html#buildincremental
+[profiles]: https://doc.rust-lang.org/cargo/reference/profiles.html
+
+If your project has not opted into enabling incremental compilation, then none of the fingerprint issues should affect your release builds.
+
+## What should a Rust programmer do in response
+
+The Internal Compiler Error asks you to report a bug, and if you can do so, we still want that information. We *want* to know about the cases that are failing.
+
+But regardless of whether or not you file a bug, the problem here can be resolved by either:
+
+ 1. deleting your incremental compilation cache (e.g. by running `cargo clean`), or
+ 2. force incremental compilation to be disabled, by setting `CARGO_INCREMENTAL=0` in your environment or `build.incremental` to `false` in the `config.toml`.
+
+We recommend that users of 1.52.0 disable incremental compilation, to avoid running into this problem.
```

Comment:
Is this language too strong? I'm reflecting on the feedback from @carols10cents [here](https://github.com/rust-lang/rust/issues/84970#issuecomment-834771043).

Should we be using the softer language of "We recommend users of 1.52.0 disable incremental compilation *if* they are encountering unstable fingerprints repeatedly" ? We still don't have a good understanding of how often people are running into this, and if they run into it repeatedly even after their first `cargo clean`...

Reply:
If it would be helpful for folks to provide you with more information about what they're seeing, this would be a great place to give more details on the most helpful information and ways to provide that. I think some folks would be happy to put up with some inconvenience if they can help to identify and fix the problems sooner.

Some folks will definitely just want to get on with their work though, and it sucks that the way to do that will make compile times slower, so I think it's important to acknowledge that. Something like:

```suggestion
To completely avoid running into this problem when using 1.52.0, we recommend disabling incremental compilation via one of the methods described above until we have had a chance to identify and fix the problems and issue a new release. We know this will make compile times slower and we know that's not an ideal workaround, but we are working to make this as temporary a situation as possible.
```